eth/tracers/api.go

```diff
@@ -226,33 +226,22 @@ func (api *API) traceChain(ctx context.Context, start, end *types.Block, config
 	}
 	sub := notifier.CreateSubscription()
 
-	// Shift the border to a block ahead in order to get the states
-	// before these blocks.
-	endBlock, err := api.blockByNumberAndHash(ctx, rpc.BlockNumber(end.NumberU64()-1), end.ParentHash())
-	if err != nil {
-		return nil, err
-	}
 	// Prepare all the states for tracing. Note this procedure can take very
 	// long time. Timeout mechanism is necessary.
 	reexec := defaultTraceReexec
 	if config != nil && config.Reexec != nil {
 		reexec = *config.Reexec
 	}
-	states, release, err := api.backend.StatesInRange(ctx, start, endBlock, reexec)
-	if err != nil {
-		return nil, err
-	}
-	defer release() // Release all the resources in the last step.
-
 	blocks := int(end.NumberU64() - start.NumberU64())
 	threads := runtime.NumCPU()
 	if threads > blocks {
 		threads = blocks
 	}
 	var (
-		pend    = new(sync.WaitGroup)
-		tasks   = make(chan *blockTraceTask, threads)
-		results = make(chan *blockTraceTask, threads)
+		pend     = new(sync.WaitGroup)
+		tasks    = make(chan *blockTraceTask, threads)
+		results  = make(chan *blockTraceTask, threads)
+		localctx = context.Background()
```

Comment:
Why do we need this local context?

Reply:
Because we are using the ctx passed by the API, but the entire tracing procedure is async and the function is returned. So the ctx is always closed.

It's the reason for `execution timeout`